Deep dive: sweeping algorithm (#103)
chhoumann/p7: report/sections/Chapter 07 - Queue System/Chapter 7.tex

In our case, there are two key variables inside the shared state structure that both endpoints need access to:
\begin{enumerate}
    \item \textbf{A sender channel:} In practice, this channel is used as a worker queue in which code submission requests are enqueued for processing. It has a corresponding \textbf{receiver channel}, which is to dequeue and schedule work. Importantly, requests sent via the sender channel are received in a first-in-first-out order.

>     \item \textbf{The job results hashmap:} Upon first receiving a code submission request, a UUID is generated. This is used as a key inside the job results hashmap, where the value is initially set to Rust \texttt{Option} value of type \texttt{TestRunnerResult} --- in other words, a key-value pair with the UUID referencing an empty job result, indicating that the work has not yet been completed. An \texttt{Option} type is simply a nullable data structure in Rust; an optional value. When the test runner has processed the code submission, the result is inserted as the value of the \texttt{Option} type. This value can then be looked up when the client requests the result processing their code submission, since the client is provided with the UUID upon sending a code submission, as previously mentioned.

Reviewer: ```suggestion
    \item \textbf{The job results hashmap:} Upon first receiving a code submission request, a UUID is generated. This is used as a key inside the job results hashmap, where the value is initially set to the Rust \texttt{Option} value of type \texttt{TestRunnerResult} --- in other words, a key-value pair with the UUID referencing an empty job result, indicating that the work has not yet been completed. An \texttt{Option} type is simply a nullable data structure in Rust; an optional value. When the test runner has processed the code submission, the result is inserted as the value of the \texttt{Option} type. This value can then be looked up when the client requests the result processing their code submission, since the client is provided with the UUID upon sending a code submission, as previously mentioned.
```

While this approach solved the initial problem of only being able to process one code submission at a time, a new problem is the fact that the hashmap continuously grows over time --- data is only ever added to the hashmap and never removed.
Consequently, the memory use of the test runner would eventually exceed its capacity, resulting in a crash.
In the following section, we introduce a sweeping system as a solution to this problem.

\section{Sweeping system}
Data is added to the job results hashmap every time a client posts a code submission.
As a reminder, this is simply a key-value pair consisting of a UUID and a Rust \texttt{Option} of type \texttt{TestRunnerResult}.
To solve the problem of the continuously growing size of the hashmap, our first idea was to remove data from the hashmap when the client polls for it.

> In other words, the data is kept in memory \textit{until} the client requests it, at which point it is returned to the client through a HTTP response and subsequently removed from the job results hashmap.

Reviewer: ```suggestion
In other words, the data is kept in memory \textit{until} the client requests it, at which point it is returned to the client through an HTTP response and subsequently removed from the job results hashmap.
```

> While this worked fine under ideal circumstances, two potential issues with this approach may arise in practice:

Reviewer: ```suggestion
While this worked fine under ideal circumstances, two potential issues with this approach arose in practice:
```

        } ~\label{line:sweep-end}~
    }).await;
}
\end{lstlisting}

Line \ref{line:last-sweep-time} initializes the last sweeping time variable as the current time, and wraps it in a mutex to prevent race conditions from arising.
Line \ref{line:sweep-config} loads the sweeping configuration data structure by reading defined values from a \texttt{.env} file using the helper function \texttt{from\_dot\_env()}.
The configurable variables include the duration between sweeps as well as the lifetime of data inside the hashmap.
Lines \ref{line:sweep-start} --- \ref{line:sweep-end} is the sweeping algorithm itself.
For each element in the hashmap where the \texttt{Option} type has a value, job results are removed if their elapsed time exceeds the configured lifetime.
This is achieved using the \texttt{retain()} function, a function that retains elements in the hashmap if and only if they fulfill some predicate.

This sweeping system achieves the desired result of removing job result data from the hashmap to prevent it from continuously growing while still retaining the data for some time, allowing clients to request the same data multiple times should an error occur.

> Furthermore, one can tweak the aforementioned environment variables to best suit the circumstances under which the system is used.

Reviewer: ```suggestion
Furthermore, one can adjust the aforementioned environment variables to best suit the circumstances under which the system is used.
```
